chia/wallet/vc_wallet/cr_cat_wallet.py

```diff
@@ -867,28 +869,41 @@ async def match_hinted_coin(self, coin: Coin, hint: bytes32) -> bool:
         This matches coins that are either CRCATs with the hint as the inner puzzle, or CRCATs in the pending approval
         state that will come to us once claimed.
         """
-        return (
+        authorized_providers_hash: bytes32 = Program.to(self.info.authorized_providers).get_tree_hash()
+        proofs_checker_hash: bytes32 = self.info.proofs_checker.as_program().get_tree_hash()
+        hint_inner_hash: bytes32 = construct_cr_layer_hash(
+            authorized_providers_hash,
+            proofs_checker_hash,
+            hint,
+        )
+        if (
             construct_cat_puzzle(
-                CAT_MOD,
+                Program.to(CAT_MOD_HASH),
                 self.info.limitations_program_hash,
-                construct_cr_layer(
-                    self.info.authorized_providers,
-                    self.info.proofs_checker.as_program(),
-                    hint,  # type: ignore
-                ),
-            ).get_tree_hash_precalc(hint)
+                hint_inner_hash,  # type: ignore
+                mod_code_hash=CAT_MOD_HASH_HASH,
+            ).get_tree_hash_precalc(hint, CAT_MOD_HASH, CAT_MOD_HASH_HASH, hint_inner_hash)
             == coin.puzzle_hash
-            or construct_cat_puzzle(
-                CAT_MOD,
+        ):
+            return True
+
+        pending_approval_inner_hash: bytes32 = construct_cr_layer_hash(
+            authorized_providers_hash,
+            proofs_checker_hash,
+            construct_pending_approval_state(hint, uint64(coin.amount)).get_tree_hash(),
+        )
+        if (
+            construct_cat_puzzle(
+                Program.to(CAT_MOD_HASH),
                 self.info.limitations_program_hash,
-                construct_cr_layer(
-                    self.info.authorized_providers,
-                    self.info.proofs_checker.as_program(),
-                    construct_pending_approval_state(hint, uint64(coin.amount)),
-                ),
-            ).get_tree_hash()
+                pending_approval_inner_hash,  # type: ignore
+                mod_code_hash=CAT_MOD_HASH_HASH,
+            ).get_tree_hash_precalc(CAT_MOD_HASH, CAT_MOD_HASH_HASH, pending_approval_inner_hash)
```

Comment:
this looks like another example of where `curry_and_treehash()` would be a lot faster. You don't actually need to construct the puzzle if all you need is the tree hash

Reply:
These ones I would like to leave because they're using the existing `construct_cat_puzzle` driver.  I've passed in all hashes here so the program that's being hashed is just:
```
(a (q . CAT_MOD_HASH) (c CAT_MOD_HASH_HASH (c TAIL_HASH (c INNER_PUZZLE_HASH 1))))
```
and therefore should not be very slow